### PR TITLE
Rework the integration between the build scan plugin and Gradle so th…

### PR DESCRIPTION
**Pull request by alkemist**
We are breaking compatibility with this release. So, we are taking the opportunity to make the user experience better if we have to break in the future. Additionally, the exchange between the plugin and build tool has been reworked to allow a better chance of safely evolving.

### CODE REVIEW

File: subprojects/core/src/main/java/org/gradle/internal/scan/config/BuildScanConfigManager.java [hunk 1]
```diff
@@ -0,0 +1,135 @@
...
+        boolean disabledViaSysProp = false;
+        boolean enabledViaSysProp = false;
+        Map<String, String> sysProps = startParameter.getSystemPropertiesArgs();
+        if (sysProps.containsKey(SYSPROP_KEY)) {
+            String sysProp = sysProps.get(SYSPROP_KEY);
+            if ("false".equals(sysProp)) {
+                disabledViaSysProp = true;
+            } else {
+                enabledViaSysProp = true;
+            }
```
Comment (etiennestuder) [inline subprojects/core/src/main/java/org/gradle/internal/scan/config/BuildScanConfigManager.java]: Not sure whether we should handle `-Dscan=foo" differently -> right now it is treated as true.

File: subprojects/core/src/main/java/org/gradle/internal/scan/config/BuildScanConfigManager.java [hunk 2]
```diff
@@ -0,0 +1,135 @@
...
+
+    BuildScanConfigManager(StartParameter startParameter, ListenerManager listenerManager, BuildScanPluginCompatibilityEnforcer compatibilityEnforcer) {
+        this.startParameter = startParameter;
+        this.listenerManager = listenerManager;
+        this.compatibilityEnforcer = compatibilityEnforcer;
+    }
+
+    @Override
+    public void init() {
+
```
Comment (etiennestuder) [inline subprojects/core/src/main/java/org/gradle/internal/scan/config/BuildScanConfigManager.java]: I'd remove that empty line.

File: subprojects/core/src/main/java/org/gradle/internal/scan/config/BuildScanConfigManager.java [hunk 3]
```diff
@@ -0,0 +1,135 @@
...
+                    );
+                }
+            }
+        });
+    }
+
+    @Override
+    public BuildScanConfig collect(BuildScanPluginMetadata pluginMetadata) {
+        if (collected) {
+            throw new IllegalStateException("configuration has already been collected");
```
Comment (etiennestuder) [inline subprojects/core/src/main/java/org/gradle/internal/scan/config/BuildScanConfigManager.java]: I'd start upper-case, i.e. 'Configuration...'

File: subprojects/core/src/main/java/org/gradle/internal/scan/config/BuildScanConfigManager.java [hunk 4]
```diff
@@ -0,0 +1,135 @@
...
+            this.configuration = configuration;
+        }
+    }
+
+    private static class Config implements BuildScanConfig {
+
+        private final boolean enabled;
+        private final boolean disabled;
+
+        Config(boolean enabled, boolean disabled) {
```
Comment (etiennestuder) [inline subprojects/core/src/main/java/org/gradle/internal/scan/config/BuildScanConfigManager.java]: Can be private

File: subprojects/core/src/test/groovy/org/gradle/initialization/DefaultGradleLauncherFactoryTest.groovy [hunk 5]
```diff
@@ -90,21 +92,33 @@ class DefaultGradleLauncherFactoryTest extends Specification {
...
+
         when:
         def launcher = factory.newInstance(startParameter, Stub(BuildRequestContext), sessionServices)
+        def c = buildScanConfig(launcher)
+
         then:
-        launcher.gradle.getServices().get(BuildScanRequest).collectDisabled()
+        !c.enabled
+        c.disabled
+    }
```
Comment (etiennestuder) [inline subprojects/core/src/test/groovy/org/gradle/initialization/DefaultGradleLauncherFactoryTest.groovy]: Should we also have a test for the case where nothing is specified explicitly? i.e. neither setBuildScan nor setNoBuildScan

File: subprojects/smoke-test/src/smokeTest/groovy/org/gradle/smoketests/BuildScanPluginSmokeTest.groovy [hunk 6]
```diff
@@ -16,33 +16,90 @@
...
+        where:
+        version << GRACEFULLY_UNSUPPORTED
     }
-}
-rootProject {
-    apply plugin: com.gradle.scan.plugin.BuildScanPlugin
-    buildScan {
-          licenseAgreementUrl = 'https://gradle.com/terms-of-service'
-          licenseAgree = 'yes'
+
```
Comment (etiennestuder) [inline subprojects/smoke-test/src/smokeTest/groovy/org/gradle/smoketests/BuildScanPluginSmokeTest.groovy]: Should we have a test for when the plugin version is 1.5 or older?

Comment (alkemist) [inline subprojects/smoke-test/src/smokeTest/groovy/org/gradle/smoketests/BuildScanPluginSmokeTest.groovy]: I don't think it's worth it. We effectively make no commitment to them working in any fashion.

**Pull request merged**